src/sentry/dynamic_sampling/tasks/common.py

```diff
@@ -345,6 +404,169 @@ def get_active_orgs_with_projects_counts(
+        self.log_state.num_iterations += 1
+        if self._enough_results_cached():
+            # we have enough in the cache to satisfy the current iteration
+            return self._get_from_cache()
+
+        if self.has_more_results:
+            # not enough for the current iteration and data still in the db top it up from db
+            query = (
+                Query(
+                    match=Entity(EntityKey.GenericOrgMetricsCounters.value),
+                    select=[
+                        Function("sum", [Column("value")], "total_count"),
+                        Column("org_id"),
+                        self.keep_count_column,
+                    ],
+                    groupby=[
+                        Column("org_id"),
+                    ],
+                    where=[
+                        Condition(
+                            # Column("timestamp"), Op.GTE, datetime.utcnow() - self.time_interval
+                            Column("timestamp"),
+                            Op.GTE,
+                            datetime.utcnow() - timedelta(hours=1),
+                        ),
+                        Condition(Column("timestamp"), Op.LT, datetime.utcnow()),
+                        Condition(Column("metric_id"), Op.EQ, self.metric_id),
+                    ],
+                    granularity=self.granularity,
+                    orderby=[
+                        OrderBy(Column("org_id"), Direction.ASC),
+                    ],
+                )
+                .set_limit(CHUNK_SIZE + 1)
+                .set_offset(self.offset)
+            )
+            request = Request(
+                dataset=Dataset.PerformanceMetrics.value, app_id="dynamic_sampling", query=query
+            )
+            self.log_state.num_db_calls += 1
+            data = raw_snql_query(
+                request,
+                referrer=Referrer.DYNAMIC_SAMPLING_COUNTERS_GET_ORG_TRANSACTION_VOLUMES.value,
+            )["data"]
+            count = len(data)
+
+            self.has_more_results = count > CHUNK_SIZE
+            self.offset += CHUNK_SIZE
+            if self.has_more_results:
+                data = data[:-1]
+            self.log_state.num_rows_total += len(data)
+            for row in data:
+                self.last_result.append(
+                    OrganizationDataVolume(
+                        org_id=row["org_id"], total=row["total_count"], indexed=row["keep_count"]
+                    )
+                )
+
+        if len(self.last_result) > 0:
+            # we have some data left return up to the max amount
+            return self._get_from_cache()  # we still have something left in cache
+        else:
+            # nothing left in the DB or cache
+            raise StopIteration()
+
+    def get_current_state(self):
+        """
+        Returns the current state of the iterator (how many orgs and projects it has iterated over)
+
+        part of the ContexIterator protocol
```

Comment:
nit: If these `get_currect_state` and `set_current_state` methods are required part of the public interface then I would expect this class, as well as the other classes with the same methods like `ContextIterator`, `TimedIterator` etc. to extend the same base class that defines that public interface.

Reply:
I prefer it as a protocol, rather than an ABC.
I get the type checking without needing to inherit ... I don't think that an interface would be 
a helpful mental model here, there is nothing to inherit here. 
I find ABCs to be quite awkward and since I don't need any runtime functionality from ABCs I don't see why we should use them here.
All I want to express is that the passed object has the required methods, a protocol is the right abstraction in my
opinion.

`ContextIterator` **is** the protocol itself ( it can't inherit). 

`TimedIterator` doesn't really need to implement the protocol (at least not yet) but I implemented it since it is trivial and makes it composable ( maybe it makes it confusing for no good reason and I should remove it).